XenModel/Utils/Helpers.cs

```diff
@@ -1844,5 +1844,10 @@ where network.Key.StartsWith(string.Format("{0}/ip", device))
                     orderby network.Key
                     select network.Value.Split(new[] { "\n", "%n" }, StringSplitOptions.RemoveEmptyEntries)).SelectMany(x => x).Distinct().ToList();
         }
+
+        public static bool GpusAvailable(IXenConnection connection)
+        {
+            return connection?.Cache?.GPU_groups?.Any(g => g.PGPUs?.Count > 0 && g.supported_VGPU_types?.Count != 0) ?? false;

```

Comment:
The only strictly necessary null check is the one on the connection. The Cache and its properties (e.g. GPU_groups) are readonly and initialised as empty. The API object properties that are ref lists (g.PGPUs, g.supported_VGPU_types) are also initialised as empty and are demarshalled by the SDK as empty if they have no values.